manifests/resource/repository/docker/proxy.pp

```diff
@@ -0,0 +1,115 @@
+#   Allow anonymous docker pull ( Docker Bearer Token Realm required ).
+# @param docker_http_port
+#   Create an HTTP connector at specified port. Normally used if the server is behind a secure proxy.
+# @param docker_https_port
+#   Create an HTTPS connector at specified port. Normally used if the server is configured for https.
+# @param docker_proxy_index_type
+#   Docker index type. See https://help.sonatype.com/repomanager3/nexus-repository-administration/formats/docker-registry/proxy-repository-for-docker#ProxyRepositoryforDocker-ConfiguringaCorrectRemoteStorageandDockerIndexURLPair
+# @param docker_proxy_index_url
+#   If docker_proxy_index_type is CUSTOM you have to set the uri of the index api.
+#
+# @example
+#   nexus::repository::docker::proxy { 'docker-docker.io':
+#      proxy_remote_url => 'https://registry-1.docker.io',
+#   }
+#
+define nexus::resource::repository::docker::proxy (
+  Stdlib::HTTPSUrl $proxy_remote_url,
+  Enum['present', 'absent'] $ensure = 'present',
+  Boolean $npm_remove_non_cataloged = false,
+  Boolean $npm_remove_quarantined = false,
+  Boolean $http_client_blocked = false,
+  Boolean $http_client_auto_block = true,
+  Boolean $negative_cache_enabled = true,
+  Integer $negative_cache_time_to_live = 1440,
+  Boolean $online = true,
+  Integer $proxy_content_max_age = 1440,
+  Integer $proxy_metadata_max_age = 1440,
+  String[1] $storage_blob_store_name = $title,
+  Boolean $storage_strict_content_type_validation = true,
+  Enum['ALLOW','ALLOW_ONCE','DENY'] $storage_write_policy = 'ALLOW',
+  Boolean $docker_v1_enabled = false,
+  Boolean $docker_force_basic_auth = true,
+  Optional[Stdlib::Port] $docker_http_port = undef,
+  Optional[Stdlib::Port] $docker_https_port = undef,
+  Enum['REGISTRY','HUB','CUSTOM'] $docker_proxy_index_type = 'HUB',
+  Optional[Stdlib::HTTPSUrl] $docker_proxy_index_url = undef,
+) {
+  nexus_repository { $title:
+    ensure     => $ensure,
+    format     => 'docker',
+    type       => 'proxy',
+    attributes => {
+      'online'          => $online,
+      'storage'         => {
+        'blobStoreName'               => $storage_blob_store_name,
+        'strictContentTypeValidation' => $storage_strict_content_type_validation,
+        'writePolicy'                 => $storage_write_policy
+      },
+      'cleanup'         => undef,
+      'proxy'           => {
+        'remoteUrl'      => $proxy_remote_url,
+        'contentMaxAge'  => $proxy_content_max_age,
+        'metadataMaxAge' => $proxy_metadata_max_age,
+      },
+      'negativeCache'   => {
+        'enabled'    => $negative_cache_enabled,
+        'timeToLive' => $negative_cache_time_to_live
+      },
+      'httpClient'      => {
+        'blocked'        => $http_client_blocked,
+        'autoBlock'      => $http_client_auto_block,
+        'connection'     => {
+          'retries'                 => undef,
+          'userAgentSuffix'         => undef,
+          'timeout'                 => undef,
```

Comment:
Why are we defining `undef` here? Are we overwriting low defaults?
Are those recommended defaults by sonatype?

Reply:
We didn't implement / use these settings yet (also not on apt/npm proxy) but if we do not have the keys the api response will differ from the definition and puppet would do api calls to change the resource on every run - which we obviously don't want.